src/embedded/embeddedSessionManager.ts

```diff
@@ -0,0 +1,134 @@
+import { v4 as uuidv4 } from 'uuid';
+import { IEmbeddedSession, IEmbeddedImpression } from '../events/types';
+import { IEmbeddedImpressionData } from './types';
+import { trackEmbeddedSession } from '../events/events';
+
+class EmbeddedSession {
+    public start?: Date
+    public end?: Date
+    public placementId?: string
+    public impressions?: Array<IEmbeddedImpression>
+    public id: string;
+
+    constructor(start?: Date, end?: Date, placementId?: string, impressions?: Array<IEmbeddedImpression>) {
+        this.start = start;
+        this.end = end;
+        this.placementId = placementId;
+        this.impressions = impressions;
+        this.id = uuidv4();
+    }
+}
+
+class EmbeddedImpressionData
+{
+    public messageId: string;
+    public displayCount: number;
+    public duration: number;
+    public start?: Date = undefined;
+
+    constructor(messageId: string, displayCount?: number, duration?: number) {
+        this.messageId = messageId;
+        this.displayCount = displayCount ? displayCount : 0;
+        this.duration = duration ? duration : 0.0;
+    }
+}
+
+export class EmbeddedSessionManager {
+    private impressions: Map<string, IEmbeddedImpression> = new Map();
+    public session: IEmbeddedSession = new EmbeddedSession(undefined, undefined, "0", undefined);
+
+    public isTracking(): Boolean {
+        return this.session.start != null
+    }
+
+    public startSession() {
+        if (this.isTracking()) {
+            console.log("Embedded session started twice");
+            return
+        }
+
+        this.session = new EmbeddedSession(new Date(), undefined, "0", undefined);
+    }
+
+    public async endSession() {
+        if (!this.isTracking()) {
+            console.log("Embedded session ended without start");
+            return
+        }
+
+        this.impressions.forEach((impressionData, messageId) => this.pauseImpression(messageId));
+        this.session.end = new Date();
+
+        if (!this.session.impressions?.length) {
+            console.log("No impressions in the session. Skipping tracking.");
+            return;
+        }
+
+        if(this.impressions.size) {
+            //reset session for next session start
+            this.session = new EmbeddedSession(undefined, undefined, "0", undefined);
```

Comment:
This `reset` seems weird to me... feels like EmbeddedSession should have it's own `reset` method on the class or something.

Reply:
It's re-initialising EmbeddedSession instance with some undefined values, i have an updated comment, in my sense there is no need of reset method separately.